### PR TITLE
Axioms (and side conditions) for tightening bounds

### PR DESCRIPTION
**Pull request by alex-ozdemir**
* Side conditions for verifying floor/ceiling-like functions
* Axioms for their correct execution
* Axioms for bound tightening.

### CODE REVIEW

File: proofs/signatures/th_lira.plf [hunk 1]
```diff
@@ -66,3 +70,119 @@
          ((>=_IntReal x y) (>=_Real (reify_int_term x) (reify_real_term y)))
          (default (fail formula))
          ))
+
+; From an arith predicate, prove the equivalent real predicate
+(declare pf_reified_arith_pred
+  (! p formula
+  (! p' formula
+    (! (th_holds p)
```
Comment (yoni206) [inline proofs/signatures/th_lira.plf]: I thought there was an identifier missing between the `!` and the `(th_holds p)` but this file type checks. How come?

File: proofs/signatures/th_lira.plf [hunk 2]
```diff
@@ -66,3 +70,119 @@
...
+; (a) k     OR
+; (b) x     OR
+; (c) k * x
+; Otherwise fails.
+; This aligns closely with the LFSCArithProof::printLinearMonomialNormalizer
+; function.
+(program negate_linear_monomial_int_term ((t (term Int))) (term Int)
+  (match t
+    ((term_int_var v) (*_Int (a_int (~ 1)) (term_int_var v)))
+    ((a_int i) (a_int (mp_neg i)))
```
Comment (yoni206) [inline proofs/signatures/th_lira.plf]: I suggest to:
1. mention that `k` is a constant and `x` is a variable
2. rename `i` to `k` (or `k` to `i`)

File: proofs/signatures/th_lira.plf [hunk 3]
```diff
@@ -66,3 +70,119 @@
...
+    (default (fail (term Int)))
+  ))
+
+; Statement that z is the greatest integer less than q.
+(declare holds_neg_of_greatest_integer_below
+  (! z mpz
+  (! q mpq
+    type)))
+
+; For proving tstatements of the above form.
```
Comment (yoni206) [inline proofs/signatures/th_lira.plf]:
```suggestion
; For proving statements of the above form.
```

File: proofs/signatures/th_lira.plf [hunk 4]
```diff
@@ -66,3 +70,119 @@
...
+(declare check_neg_of_greatest_integer_below
+  (! z mpz
+  (! q mpq
+    (! sc_check (^ (is_greatest_integer_below (mp_neg z) q) tt)
+       (holds_neg_of_greatest_integer_below z q)))))
+
+; Axiom for tightening [Int] < q.
+(declare tighten_not_>=_IntReal
+  (! t       (term Int)  ; Omit
```
Comment (yoni206) [inline proofs/signatures/th_lira.plf]: What are all these `; Omit`s?

File: proofs/signatures/th_lira.plf [hunk 5]
```diff
@@ -66,3 +70,119 @@
...
+(program negate_linear_int_term ((t (term Int))) (term Int)
+  (match t
+    ((term_int_var v) (negate_linear_monomial_int_term t))
+    ((a_int i) (negate_linear_monomial_int_term t))
+    ((+_Int x y) (+_Int (negate_linear_int_term x) (negate_linear_int_term y)))
+    ((*_Int x y) (negate_linear_monomial_int_term t))
+    (default (fail (term Int)))
+  ))
+
+; Statement that z is the greatest integer less than q.
```
Comment (yoni206) [inline proofs/signatures/th_lira.plf]:
```suggestion
; Statement that z is the negation of greatest integer less than q.
```
?

File: proofs/signatures/th_lira.plf [hunk 6]
```diff
@@ -66,3 +70,119 @@
...
+    type)))
+
+; For proving tstatements of the above form.
+(declare check_neg_of_greatest_integer_below
+  (! z mpz
+  (! q mpq
+    (! sc_check (^ (is_greatest_integer_below (mp_neg z) q) tt)
+       (holds_neg_of_greatest_integer_below z q)))))
+
+; Axiom for tightening [Int] < q.
```
Comment (yoni206) [inline proofs/signatures/th_lira.plf]: Maybe write what is the tightened constraint (is it `-[Int]>=-q` ?)

File: proofs/signatures/th_lira.plf [hunk 7]
```diff
@@ -66,3 +70,119 @@
...
+  (! t       (term Int)  ; Omit
+  (! neg_t   (term Int)  ; Omit
+  (! real_bound mpq    ; Omit
+  (! neg_int_bound mpz ; Omit
+    (! pf_step (holds_neg_of_greatest_integer_below neg_int_bound real_bound)
+    (! pf_real_bound (th_holds (not (>=_IntReal t (a_real real_bound))))
+      (! sc_neg (^ (negate_linear_int_term t) neg_t)
+        (th_holds (>=_IntReal neg_t (term_int_to_real (a_int neg_int_bound))))))))))))
+
+; Axiom for tightening [Int] >= q.
```
Comment (yoni206) [inline proofs/signatures/th_lira.plf]: Maybe write what is the tightened constraint?

File: proofs/signatures/th_lira.plf [hunk 8]
```diff
@@ -66,3 +70,119 @@
...
+    type)))
+
+; For proving statements of the above form.
+(declare check_neg_of_greatest_integer_below_int
+  (! z mpz
+  (! z' mpz
+    (! sc_check (^ (is_greatest_integer_below (mp_neg z) (mpz_to_mpq z')) tt)
+       (holds_neg_of_greatest_integer_below_int z z')))))
+
+; Axiom for tightening [Int] < i.
```
Comment (yoni206) [inline proofs/signatures/th_lira.plf]: Maybe write what is the tightened constraint?

File: proofs/signatures/th_lira.plf (hunk 1 above)
Comment (alex-ozdemir) [reply proofs/signatures/th_lira.plf]: Interesting! Perhaps LFSC regards those identifiers as optional? I'm actually not sure.

Comment (alex-ozdemir) [reply proofs/signatures/th_lira.plf]: I'll put the identifier in there!

File: proofs/signatures/th_lira.plf (hunk 2 above)
Comment (alex-ozdemir) [reply proofs/signatures/th_lira.plf]: Good idea!

File: proofs/signatures/th_lira.plf (hunk 4 above)
Comment (alex-ozdemir) [reply proofs/signatures/th_lira.plf]: To remind myself that these can be inferred from the other arguments, and they should be omitted (replaced with an underscore) when invoking the axiom.

I'll put a note at the top suggesting this.

File: proofs/signatures/th_lira.plf (hunk 6 above)
Comment (alex-ozdemir) [reply proofs/signatures/th_lira.plf]: Good idea!

File: proofs/signatures/th_lira.plf (hunk 7 above)
Comment (alex-ozdemir) [reply proofs/signatures/th_lira.plf]: Will do! It's `[Int] >= ceil(q)`

File: proofs/signatures/th_lira.plf (hunk 8 above)
Comment (alex-ozdemir) [reply proofs/signatures/th_lira.plf]: Will do! It's `[Int] >= -(i - 1)`.

**Comment by alex-ozdemir** [comment]
Thanks for the comments, Yoni.

Documenting the tightened inequalities is a great idea, and I've gone ahead an done so.

I also believe that I've addressed you other comments. Let me know if you have more concerns!

File: proofs/signatures/th_lira.plf (hunk 1 above)
Comment (yoni206) [reply proofs/signatures/th_lira.plf]: Sounds good. Thanks.
Maybe @ajreynol can tell us what's going on with missing identifiers type-checking in lfsc? :)

**Pull request merged**